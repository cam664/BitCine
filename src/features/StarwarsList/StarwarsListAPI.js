import fetch from 'isomorphic-fetch';

export const fetchPeopleAPI = () => {
  return fetch(`https://swapi.co/api/people/`);
}