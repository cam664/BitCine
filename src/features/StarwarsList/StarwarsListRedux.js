import { 
  fetchPeopleAPI, 
  fetchShipsAPI, 
  fetchPeopleNextAPI, 
  fetchShipsNextAPI 
} from './StarwarsListAPI';

export const fetchListActions = {
  REQUEST: 'StarwarsList/fetchList/REQUEST',
  SUCCESS: 'StarwarsList/fetchList/SUCCESS',
  FAILURE: 'StarwarsList/fetchList/FAILURE'
}

const fetchShipsActions = {
  REQUEST: 'StarwarsList/fetchShips/REQUEST',
  SUCCESS: 'StarwarsList/fetchShips/SUCCESS',
  FAILURE: 'StarwarsList/fetchShips/FAILURE'
}

const fetchListNextActions = {
  REQUEST: 'StarwarsList/fetchListNext/REQUEST',
  SUCCESS: 'StarwarsList/fetchListNext/SUCCESS',
  FAILURE: 'StarwarsList/fetchListNext/FAILURE'
}

const fetchShipsNextActions = {
  REQUEST: 'StarwarsList/fetchShipsNext/REQUEST',
  SUCCESS: 'StarwarsList/fetchShipsNext/SUCCESS',
  FAILURE: 'StarwarsList/fetchShipsNext/FAILURE'
}

const initialState = {
  list: {
    status: 'loading',
    data: [],
    next: null,
    spliceIdx: 7
  },
  ships: {
    status: 'loading',
    data: [],
    next: null
  }
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    // fetchList
    case fetchListActions.REQUEST:
      return {
        ...state,
        list: {
          ...state.list,
          status: 'loading',
          data: [],
        }
      }
    case fetchListActions.SUCCESS:
      return {
        ...state,
        list: {
          ...state.list,
          status: 'success',
          data: action.data,
          next: action.next,
          spliceIdx: action.spliceIdx
        }
      }
    case fetchListActions.FAILURE:
      return {
        ...state,
        list: {
          ...state.list,
          status: 'failure',
          data: [],
        }
      }
      // fetchShips
      case fetchShipsActions.REQUEST:
        return {
          ...state,
          ships: {
            ...state.ships,
            status: 'loading',
            data: [],
            
          }
        }
      case fetchShipsActions.SUCCESS:
        return {
          ...state,
          ships: {
            ...state.ships,
            status: 'success',
            data: action.data,
            next: action.next,
            
          }
        }
      case fetchShipsActions.FAILURE:
        return {
          ...state,
          ships: {
            ...state.ships,
            status: 'failure',
            data: [],
            
          }
        }
      // fetchListNext
      case fetchListNextActions.REQUEST:
        return {
          ...state,
          list: {
            ...state.list,
            status: 'loading', 
          }
        }
      case fetchListNextActions.SUCCESS:
        return {
          ...state,
          list: {
            ...state.list,
            status: 'success',
            data: action.data,
            next: action.next,
          }
        }
      case fetchListNextActions.FAILURE:
        return {
          ...state,
          list: {
            ...state.list,
            status: 'failure', 
          }
        }
      // fetchShipsNext
      case fetchShipsNextActions.REQUEST:
        return {
          ...state,
          ships: {
            ...state.ships,
            status: 'loading', 
          }
        }
      case fetchShipsNextActions.SUCCESS:
        return {
          ...state,
          ships: {
            ...state.ships,
            status: 'success',
            data: action.data,
            next: action.next,
          }
        }
      case fetchShipsNextActions.FAILURE:
        return {
          ...state,
          ships: {
            ...state.ships,
            status: 'failure', 
          }
        }

      default:
        return state;
  }
}

export function fetchList() {
  return async (dispatch, getState) => {
    dispatch({ type: fetchListActions.REQUEST });
    dispatch({ type: fetchShipsActions.REQUEST });

    try {
      const res = await Promise.all([
        fetchPeopleAPI().then(res => res.json()), 
        fetchShipsAPI().then(res => res.json())
      ]);
      
      const currState = getState();
      let { spliceIdx } = currState.StarwarsList.list;

      let peopleRes = res[0];
      let shipsRes = res[1]
      let listPayload = [...peopleRes.results];
      let shipsPayload = [...shipsRes.results];

      listPayload.splice(spliceIdx, 0, shipsPayload.shift())
      let nextSpliceIdx = spliceIdx + 8; 
      
      dispatch({ 
        type: fetchListActions.SUCCESS,
        data: listPayload,
        next: peopleRes.next,
        spliceIdx: nextSpliceIdx
      });
      dispatch({
        type: fetchShipsActions.SUCCESS,
        data: shipsPayload,
        next: shipsRes.next
      });
    } catch(e) {
      dispatch({ type: fetchListActions.FAILURE });
      dispatch({ type: fetchShipsActions.FAILURE });
    }
  }
}

export function fetchListNext() {
  return async (dispatch, getState) => {
    const currState = getState();
    const listNextUrl = currState.StarwarsList.list.next;
    const ships = currState.StarwarsList.ships.data;

    let dataReqs = [fetchPeopleNextAPI(listNextUrl)];
    if (ships.length < 2) {
      dataReqs.push(fetchShipsNextAPI(currState.StarwarsList.ships.next))
    }

    if (dataReqs.length === 1) {
      dispatch({ type: fetchListNextActions.REQUEST });
    } else {
      dispatch({ type: fetchListNextActions.REQUEST });
      dispatch({ type: fetchShipsNextActions.REQUEST });
    }

    try {
      const res = await Promise.all(dataReqs)
        .then(res => Promise.all(res.map(res => res.json())));

      const { spliceIdx } = currState.StarwarsList.list;
      let currList = currState.StarwarsList.list.data;

      let peopleRes = res[0];
      let shipsPayload = res[1] ? [...ships, ...res[1].results] : ships;
      let shipsNext = res[1] ? res[1].next : currState.StarwarsList.ships.next;
      let listPayload = [...currList, ...peopleRes.results];

      listPayload.splice(spliceIdx, 0, shipsPayload.shift())
      let nextSpliceIdx = spliceIdx + 8; 

      while (nextSpliceIdx < listPayload.length) {
        listPayload.splice(nextSpliceIdx, 0, shipsPayload.shift())
        nextSpliceIdx += 8;
      }
      
      dispatch({ 
        type: fetchListActions.SUCCESS,
        data: listPayload,
        next: peopleRes.next,
        spliceIdx: nextSpliceIdx
      });
      dispatch({
        type: fetchShipsActions.SUCCESS,
        data: shipsPayload,
        next: shipsNext
      });
    } catch(e) {
      dispatch({ type: fetchListNextActions.FAILURE });
      dispatch({ type: fetchShipsNextActions.FAILURE });
    }
  }
}