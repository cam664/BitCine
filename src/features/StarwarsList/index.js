export { default } from './StarwarsList';

export { fetchPeopleAPI } from './StarwarsListAPI';

export { default as StarwarsListReducer } from './StarwarsListRedux';

export { fetchPeople } from './StarwarsListRedux';