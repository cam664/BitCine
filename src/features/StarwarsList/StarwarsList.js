import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import Table from '../../common/Table';
import Spinner from '../../common/Spinner';

import { fetchList, fetchListNext } from './StarwarsListRedux';

const columns = [
  {
    label: 'Name',
    field: 'name',
    width: '40%'
  },
  {
    label: 'Birth Year',
    field: 'birth_year',
    width: '20%'
  },
  {
    label: 'Height',
    field: 'height',
    width: '20%',
  },
  {
    label: 'Mass',
    field: 'mass',
    width: '20%'
  },
]

let canFetchFlag = true;

const StarwarsList = () => {
  const dispatch = useDispatch();

  const [isLoadingFromClient, setIsLoadingFromClient] = useState(false);
  const listData = useSelector(state => state.StarwarsList);
  useEffect(() => { 
    async function fetchData() {
      if (listData.list.status !== 'success') {
        setIsLoadingFromClient(() => true);
        await dispatch(fetchList());
        setIsLoadingFromClient(() => false);
      }
    }
    fetchData();
  }, [])
  
  const [listIsLoading, setListLoading] = useState(false);
  const onMore = async () => {
    if (canFetchFlag) {
      canFetchFlag = false;

      if (!listData.list.next) return;

      setListLoading(() => true);
      await dispatch(fetchListNext())
      setListLoading(() => false);
      canFetchFlag = true;
    }
  }

  return (
    <>
      {(listData.list.status !== 'success' && isLoadingFromClient) 
      ? (
        <Spinner />
      ) : (
        <Table columns={columns} data={listData.list.data} onMore={onMore} isLoading={listIsLoading} />
      )}
    </>
  );
}

StarwarsList.serverFetch = fetchList;

export default StarwarsList;