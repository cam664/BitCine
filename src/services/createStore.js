import { createStore, combineReducers, applyMiddleware } from "redux";
import thunkMiddleware from "redux-thunk";

import { StarwarsListReducer } from '../features/StarwarsList';

const reducer = combineReducers({
  StarwarsList: StarwarsListReducer,
});

export default (initialState) => 
  createStore(
    reducer, 
    initialState, 
    applyMiddleware(thunkMiddleware)
  );
