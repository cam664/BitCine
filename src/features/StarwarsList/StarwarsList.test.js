import React from 'react';
import renderer from 'react-test-renderer';

import { Provider as ReduxProvider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import StarwarsList from './StarwarsList';
import { default as StarwarsListReducer } from './StarwarsListRedux';
import { fetchListActions } from './StarwarsListRedux';

import createStore from '../../services/createStore';

const store = createStore(window.__PRELOADED_STATE__);

describe('StarwarsList', () => {
  it('renders a snapshot', () => {
    const component = renderer.create(
      <ReduxProvider store={store}>
        <Router>
          <StarwarsList />
        </Router>
      </ReduxProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('reducer', () => {
    const payload = { 
      status: 'success',
      data: [{name: 'John Smith'}, {name: 'Jane Doe'}],
      next: 'https://url.com',
      spliceIdx: 7
    };

    it('should return a data object with correct shape', () => {
      const state = { 
        list: {
          status: 'loading',
          data: [],
          next: null,
          spliceIdx: 7
        },
        ships: {
          status: 'loading',
          data: [],
          next: null
        }
      };
      const newState = StarwarsListReducer(state, {
        type: fetchListActions.SUCCESS,
        data: payload.data,
        next: payload.next,
        spliceIdx: payload.spliceIdx
      });
      expect(newState).toEqual({
        list: {
          data: payload.data,
          next: payload.next,
          spliceIdx: 7,
          status: 'success'
        },
        ships: {
          data: [],
          next: null,
          status: 'loading'
        }
      });
    });
  });
});