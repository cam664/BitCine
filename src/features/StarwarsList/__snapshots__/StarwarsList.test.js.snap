// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarwarsList renders a snapshot 1`] = `
<section
  className="css-1rt94nh-Table"
>
  <table>
    <thead>
      <tr>
        <th
          style={
            Object {
              "width": "40%",
            }
          }
        >
          <div>
            Name
          </div>
        </th>
        <th
          style={
            Object {
              "width": "20%",
            }
          }
        >
          <div>
            Birth Year
          </div>
        </th>
        <th
          style={
            Object {
              "width": "20%",
            }
          }
        >
          <div>
            Height
          </div>
        </th>
        <th
          style={
            Object {
              "width": "20%",
            }
          }
        >
          <div>
            Mass
          </div>
        </th>
      </tr>
    </thead>
  </table>
  <div
    className="css-1mzwwa2-Table"
    onScroll={[Function]}
  >
    <table>
      <tbody />
    </table>
  </div>
</section>
`;
